Coupled-Electron-Pair-Approximation/OLCCD.py

```diff
@@ -0,0 +1,138 @@
+"""
+Computes the Orbital-Optimized LCCD correlation energy.
+Equations taken from [Bozkaya:2013:054104].
+
+__authors__   =  "Jonathon P. Misiewicz"
+__credits__   =  ["Jonathon P. Misiewicz"]
+
+__copyright__ = "(c) 2014-2020, The Psi4NumPy Developers"
+__license__   = "BSD-3-Clause"
+"""
+
+import numpy as np
+import psi4
+from scipy import linalg as spla
+from DSD import DirectSumDiis
+from integrals import integrals
+
+### Settings
+
+mol = psi4.geometry("""
+O
+H 1 1.0
+H 1 1.0 2 104.5
+symmetry c1""")
+scf_type = "pk"
+target_convergence = 7
+maxiter = 50
+compare_psi4 = True
+basis = "cc-pvdz"
+
+### Setup
+psi4.set_options({"scf_type": scf_type, "e_convergence": target_convergence + 1, "basis": basis})
+I, F, intermed = integrals(mol, singles=True, return_intermediates=True)
+t1 = np.zeros(F["ov"].shape)
+t2 = np.zeros(I["oovv"].shape)
+dsd = DirectSumDiis(3, 8)
+num_occ, num_vir = t1.shape
+
+### Main Loop
+for i in range(1, maxiter + 1):
+
+    if i != 1:
+        # Compute new orbitals and transform into them. See Section IIA5 of Bozkaya.
+        Zoo = np.zeros((num_occ, num_occ))
+        Zvv = np.zeros((num_vir, num_vir))
+        X = np.block([[Zoo, -t1],
+                     [t1.T, Zvv]])
+        U = spla.expm(X)
+        C = np.hstack((intermed["O"], intermed["V"]))
+        C = C @ U
+        C_O, C_V = np.hsplit(C, [num_occ])
+        I = {
+            "oovv": np.einsum('pP,qQ,rR,sS,pqrs->PQRS', C_O, C_O, C_V, C_V, intermed["TEI"], optimize = True),
+            "oooo": np.einsum('pP,qQ,rR,sS,pqrs->PQRS', C_O, C_O, C_O, C_O, intermed["TEI"], optimize = True),
+            "voov": np.einsum('pP,qQ,rR,sS,pqrs->PQRS', C_V, C_O, C_O, C_V, intermed["TEI"], optimize = True),
+            "vvvv": np.einsum('pP,qQ,rR,sS,pqrs->PQRS', C_V, C_V, C_V, C_V, intermed["TEI"], optimize = True),
+            "ovvv": np.einsum('pP,qQ,rR,sS,pqrs->PQRS', C_O, C_V, C_V, C_V, intermed["TEI"], optimize = True),
+            "ooov": np.einsum('pP,qQ,rR,sS,pqrs->PQRS', C_O, C_O, C_O, C_V, intermed["TEI"], optimize = True)
+            }
+        H = {
+            "oo": np.einsum('pP, qQ, pq -> PQ', C_O, C_O, intermed["OEI"], optimize = True),
+            "ov": np.einsum('pP, qQ, pq -> PQ', C_O, C_V, intermed["OEI"], optimize = True),
+            "vv": np.einsum('pP, qQ, pq -> PQ', C_V, C_V, intermed["OEI"], optimize = True)
+            }
+        F = {
+            "oo": H["oo"] + np.einsum('iP iQ -> PQ', I["oooo"], optimize = True),
+            "ov": H["ov"] + np.einsum('iP iQ -> PQ', I["ooov"], optimize = True),
+            "vv": H["vv"] - np.einsum('Pi iQ -> PQ', I["voov"], optimize = True)
+            }
+    else:
```

Comment:
It might be cleaner if these are initialized before the iteration loop so that we don't need the else.